packages/cdktf/lib/validations/validate-provider-presence.ts

```diff
@@ -0,0 +1,50 @@
+import { IConstruct, IValidation, Node } from "constructs";
+import { TerraformProvider } from "../terraform-provider";
+import { TerraformResource } from "../terraform-resource";
+import { TerraformDataSource } from "../terraform-data-source";
+
+export class ValidateProviderPresence implements IValidation {
+  public providerNames: Set<string> = new Set();
+  public foundProviders: TerraformProvider[] = [];
+
+  constructor(protected host: IConstruct) {}
+
+  public check(node: IConstruct) {
+    if (
+      node instanceof TerraformResource ||
+      node instanceof TerraformDataSource
+    ) {
+      if (node.terraformGeneratorMetadata) {
+        this.providerNames.add(node.terraformGeneratorMetadata.providerName);
+      }
+    }
+
+    if (node instanceof TerraformProvider) {
+      this.foundProviders.push(node);
+    }
+
+    for (const child of Node.of(node).children) {
+      this.check(child);
+    }
+  }
+
+  public validate() {
+    this.check(this.host);
+
+    const missingProviders = Array.from(this.providerNames).filter((name) => {
+      return !this.foundProviders.some(
+        (p) => p.terraformGeneratorMetadata?.providerName === name
+      );
+    });
+
+    if (missingProviders.length === 0) {
+      return [];
+    } else {
+      return [
+        `Found resources without a matching povider. Please make sure to add the following providers to your stack: ${missingProviders.join(
```

Comment:
Perhaps in the future we could exclude providers that don't require any configuration.

Reply:
I'm sure I ran into an issue where I forgot to initialized the docker provider and the program errored. The providers also help to set the right name in the required_providers (org is defaulted to hashicorp).

Reply:
I think that is true, but I don't recall that being the case in hcl (maybe it changed in newer versions).

Reply:
> Perhaps in the future we could exclude providers that don't require any configuration.

Do you know if that's documented somewhere?

Reply:
Everything in the schema would be optional. Post Terraform 0.13, probably only works for providers in the `hashicorp` namespace.